server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/ProjectStatusAction.java

```diff
@@ -0,0 +1,178 @@
+    ProjectStatusWsRequest projectStatusWsRequest = toProjectStatusWsRequest(request);
+    ProjectStatusWsResponse projectStatusWsResponse = doHandle(projectStatusWsRequest);
+
+    writeProtobuf(projectStatusWsResponse, request, response);
+  }
+
+  private ProjectStatusWsResponse doHandle(ProjectStatusWsRequest request) {
+    checkScanOrAdminPermission();
+
+    DbSession dbSession = dbClient.openSession(false);
+    try {
+      String taskUuid = request.getAnalysisId();
+      CeActivityDto ceActivityDto = getCeActivity(dbSession, taskUuid);
+      SnapshotDto snapshotDto = getSnapshot(dbSession, ceActivityDto.getSnapshotId(), taskUuid);
+      MeasureDto measureDto = getQualityGateDetailsMeasure(dbSession, snapshotDto, taskUuid);
+
+      return buildResponse(measureDto.getData(), taskUuid);
+    } finally {
+      dbClient.closeSession(dbSession);
+    }
+  }
+
+  private CeActivityDto getCeActivity(DbSession dbSession, String taskUuid) {
+    return checkFoundWithOptional(dbClient.ceActivityDao().selectByUuid(dbSession, taskUuid),
+      "Task with id '%s' is not found", taskUuid);
+  }
+
+  private MeasureDto getQualityGateDetailsMeasure(DbSession dbSession, SnapshotDto snapshotDto, String taskUuid) {
+    List<MeasureDto> measures = dbClient.measureDao().selectBySnapshotIdAndMetricKeys(snapshotDto.getId(),
+      Collections.singleton(CoreMetrics.QUALITY_GATE_DETAILS_KEY), dbSession);
+    if (measures.isEmpty()) {
+      throw new NotFoundException(format("Quality gate details of task with id '%s' are not found", taskUuid));
+    }
+
+    return measures.get(0);
+  }
+
+  private SnapshotDto getSnapshot(DbSession dbSession, @Nullable Long snapshotId, String taskId) {
+    SnapshotDto snapshotDto = (snapshotId == null)
+      ? null
+      : dbClient.snapshotDao().selectById(dbSession, snapshotId);
+
+    return checkFound(snapshotDto, "Analysis for task id '%s' is not found", taskId);
+  }
+
+  private static ProjectStatusWsResponse buildResponse(String measureData, String taskUuid) {
+    ProjectStatus.Builder responseBuilder = ProjectStatus.newBuilder();
+    try {
+      JsonParser parser = new JsonParser();
+      JsonObject json = parser.parse(measureData).getAsJsonObject();
+
+      String status = measureLevelToQualityGateStatus(json.get("level").getAsString());
+      responseBuilder.setStatus(status);
+      JsonArray conditions = json.getAsJsonArray("conditions");
+      ProjectStatusWsResponse.Condition.Builder conditionBuilder = ProjectStatusWsResponse.Condition.newBuilder();
+      Gson gson = GsonHelper.create();
+      for (JsonElement condition : conditions) {
+        responseBuilder.addConditions(conditionBuilder
+          .clear()
+          .putAllCondition(gson.<Map<String, String>>fromJson(condition, new MapJsonType().getType())));
+      }
+      return ProjectStatusWsResponse.newBuilder().setProjectStatus(responseBuilder).build();
+    } catch (Exception e) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1166)